genienlp/ned/bootleg.py

```diff
@@ -323,103 +215,70 @@ def collect_features_per_line(self, line, threshold):
-                examples[n].question = self.add_type_tokens(ex.question, ex.question_feature)
-
-            else:
-                # context is the utterance field
-                for i in range(len(tokens_type_ids)):
-                    examples[n].context_feature[i].type_id = tokens_type_ids[i]
-                    examples[n].context_feature[i].type_prob = tokens_type_probs[i]
-                    examples[n].context_feature[i].qid = tokens_qids[i]
-                examples[n].context = self.add_type_tokens(ex.context, ex.context_feature)
-
-    def merge_embeds(self, file_list):
-        all_emb_data = []
-        for file_name in file_list:
-            emb_file = f'{self.args.bootleg_output_dir}/{file_name}_bootleg/{self.ckpt_name}/bootleg_embs.npy'
-            with open(emb_file, 'rb') as fin:
-                emb_data = np.load(fin)
-                all_emb_data.append(emb_data)
-
-        all_emb_data = np.concatenate(all_emb_data, axis=0)
-
-        # add embeddings for the padding and unknown special tokens
-        new_emb = np.concatenate([np.zeros([2, all_emb_data.shape[1]], dtype='float'), all_emb_data], axis=0)
-
-        os.makedirs(f'{self.args.bootleg_output_dir}/bootleg/eval/{self.ckpt_name}', exist_ok=True)
-        np.save(f'{self.args.bootleg_output_dir}/bootleg/eval/{self.ckpt_name}/ent_embedding.npy', new_emb)
-
-
-class BootlegAnnotator(object):
+                all_token_type_ids.append(tokens_type_ids)
+                all_token_type_probs.append(tokens_type_probs)
+                all_token_qids.append(tokens_qids)
+
+        all_token_type_ids = all_token_type_ids[: self.args.subsample]
+        all_token_type_probs = all_token_type_probs[: self.args.subsample]
+        all_token_qids = all_token_qids[: self.args.subsample]
+
+        self.replace_features_inplace(examples, all_token_type_ids, all_token_type_probs, all_token_qids, utterance_field)
+
+    def dump_entities_with_labels(self, examples, path, utterance_field):
+        input_file_dir = os.path.dirname(path)
+        input_file_name = os.path.basename(path.rsplit('.', 1)[0] + '_bootleg.jsonl')
+        data_overrides = ["--data_config.data_dir", input_file_dir, "--data_config.test_dataset.file", input_file_name]
+
+        # get config args
+        config_overrides = self.fixed_overrides
+        config_overrides.extend(data_overrides)
+        config_args = self.create_config(config_overrides)
+
+        # create jsonl files from input examples
+        # jsonl is the input format bootleg expects
+        self.create_jsonl(path, examples, utterance_field)
+
+        # extract mentions and mention spans in the sentence and write them to output jsonl files
+        self.extract_mentions(path)
+
+        # find the right entity candidate for each mention
+        self.disambiguate_mentions(config_args)
+
+
+class ServingBootlegEntityDisambiguator(BatchBootlegEntityDisambiguator):
```

Comment:
Do you need this inheritance? What are you inheriting?

Reply:
We discussed this on slack. ServingBootlegEntityDisambiguator uses several methods and attributes from BatchBootlegEntityDisambiguator. Turning them into utility functions is not viable since methods are called within each other.